src/main/java/org/datadog/jmxfetch/Instance.java

```diff
@@ -256,6 +254,32 @@ public Instance(
         }
     }
 
+    /** this function is unsafe, please call holding a lock. */
+    public synchronized void updateServiceList(List<String> services) {
+        this.services = services;
```

Comment:
I might add a null check here, not that we would try to hand a null list but it's probably best to ensure a list with some default service name at the least

Reply:
Possibly even make an unmodifiable copy here too

Reply:
@devinsba the service is not a guaranteed tag in the jmxfetch world, so we have to be ready for a null list. The rest of the code should be resilient to handling `Instance.services` being null.

Finally, please correct me if I'm wrong - replacing a reference variable is not an atomic operation, so if we're going to be accessing `this.services` concurrently wouldn't we need to have some synchronization? If we don't I'm happy to remove the synchronization, but let's keep in mind the scenario where we call this function somehow from the tracer at some random time to update the service list.

Reply:
Yes. Reference writes are atomic https://docs.oracle.com/javase/tutorial/essential/concurrency/atomic.html

> Reads and writes are atomic for reference variables and for most primitive variables (all types except long and double).

In this case the synchronization wouldn't accomplish the goal I don't think since synchronized methods all have their own lock. They don't share